raster-foundry/raster-foundry
# Add band count validation to local file upload modal

## Overview

For each file being uploaded, the modal uses [Loam](https://www.npmjs.com/package/loam) and [gdal-js](https://www.npmjs.com/package/gdal-js) to validate that the file has the same number of bands as the selected datasource. If not, an alert icon with a tooltip explaining the problem is displayed.

### Checklist


### Demo

![bandvalidation2](https://user-images.githubusercontent.com/447977/36978661-14ed7d28-2053-11e8-80e5-4a7445efda04.png)

### Notes

I've been running using the staging API, which (I think) prevented the upload from ultimately working, so I haven't been able to test that part fully. I wanted to get this PR up though so that @alkamin can use it as a reference this week if needed.

Although I think this offers useful functionality on its own, there are a number of things that could be done to improve it; I'm happy to do some or all of them, and I'll leave it up to you all to decide which ones are important right now.

1. ~Error-handling is basically nonexistent. If you upload a GeoTiff that is invalid or otherwise not a GeoTiff (e.g. because it's an Excel spreadsheet), it will simply fail to show up in the list. There is a quick fix for this, which is to check whether the dataset pointer returned by Loam is 0 (null), but the real fix (which is, not surprisingly, quite a bit more work) is for Loam to hook into GDAL's error-handling code and reject the promise if an error is detected.~ Better error handling has been added, see below.
2. Firefox currently issues a console warning about the mimetype for the GDAL WebAssembly being incorrect. This doesn't seem to hurt anything, but it does cause Firefox to download the WebAssembly _twice_. That's not a big deal on local development but it could be more of an issue in production since the zipped WebAssembly is about 1.5MB.
3. Building on the previous point, this PR doesn't really address what deployment should look like for Loam and GDAL. Since they add files that can't be part of the existing bundles, I needed to adjust our existing Webpack config to get things working in development. I assume that we'll also need to make changes to the static asset deployment pipeline as well, but I don't really know much about how that works.
4. AngularJS promises and native promises don't mesh as well as one might like. I've gotten that working where it matters here, but there are a few spots (mainly in the cleanup code) where I've ignored the issue because we're not using the result of the promise anywhere. Before using Loam elsewhere in the code, or expanding its usage for the upload modal, it'd be good to break it out into a service that handles converting the promises internally so that nothing else has to worry about that complexity.
5. A band-count mismatch doesn't prevent the user from uploading the file. That seemed like a reasonable approach to start, but we may want to be stricter about it.

## Testing Instructions

 * `yarn install` from `app-frontend` to get latest dependencies
 * `./scripts/server` and `yarn run start` (and anything else necessary to get a local dev environment going; mine is old and not working anymore, so I'm not sure what the steps are exactly).
 * Bring up the data import interface. Choose a Datasource that has a number of bands specified in the database (a lot of mine didn't).
 * Choose a local file upload import type, and then select some GeoTiffs on your local machine. Make sure that some of them match the number of bands in the Datasource, and others don't.
 * Verify that the tiffs that have the same number of bands as the Datasource don't display any warnings, and the tiffs that _don't_ have the same number of bands display a warning.
 * Verify that the warning messages correctly display the number of bands in the geotiff and Datasource.
 * Verify that other modal functionality works (add, delete, etc.) and that you can upload the tiffs successfully.
